### PR TITLE
Added JAX for PyTorch users tutorial

### PR DESCRIPTION
**Pull request by vfdev-5**

### CODE REVIEW

File: docs/JAX_for_PyTorch_users.md [hunk 1]
```diff
@@ -0,0 +1,890 @@
...
+rand_tensor2 = jax.random.uniform(key, (2, 3))
+assert (rand_tensor1 == rand_tensor2).all()
+
+k1, k2 = jax.random.split(key, num=2)
+rand_tensor1 = jax.random.uniform(k1, (2, 3))
+rand_tensor2 = jax.random.uniform(k2, (2, 3))
+assert (rand_tensor1 != rand_tensor2).all()
+```
+
+For further discussion on random number in NumPy and JAX check [this tutorial](https://jax.readthedocs.io/en/latest/random-numbers.html).
```
Comment (jakevdp) [inline docs/JAX_for_PyTorch_users.md:142]:
```suggestion
For further discussion on random numbers in NumPy and JAX check [this tutorial](https://jax.readthedocs.io/en/latest/random-numbers.html).
```

File: docs/JAX_for_PyTorch_users.md [hunk 2]
```diff
@@ -0,0 +1,890 @@
...
+
+x_torch = torch.zeros(3, 4)
+x_jax = jnp.zeros((3, 4))
+print(f"Default float dtypes, PyTorch: {x_torch.dtype} and Jax: {x_jax.dtype}")
+print(f"Default devices, PyTorch: {x_torch.device} and Jax: {x_jax.device}")
+```
+
++++ {"id": "1o1u2N1lL7I3"}
+
+Note that this tutorial does not cover the API to set values on indices on an existing JAX Array. For more details, please check the corresponding JAX documentation: https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.ndarray.at.html.
```
Comment (jakevdp) [inline docs/JAX_for_PyTorch_users.md:225]:
```suggestion
For some discussion of JAX's alternative to in-place mutation, refer to https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.ndarray.at.html.
```

File: docs/JAX_for_PyTorch_users.md [hunk 3]
```diff
@@ -0,0 +1,890 @@
...
+---
+# create an array on CPU and check the device
+x_cpu = jnp.ones((3, 4), device=cpu_device)
+print(x_cpu.device, )
+# create an array on GPU
+x_gpu = jnp.ones((3, 4), device=cuda_device)
+print(x_gpu.device)
+```
+
+In JAX it is possible to implicitly move an array from one device to another when we operate with arrays on different devices. Namely, when we create a JAX Array **without explicitly** specifying a device, unlike PyTorch, in the computation it can be implicitly moved to the explicitly specified device of another JAX Array.
```
Comment (jakevdp) [inline docs/JAX_for_PyTorch_users.md:266]:
```suggestion
In PyTorch we are used to device placement always being explicit. JAX can operate this way via explicit device placement as above, but unless the device is specified the array will remain *uncommitted*: i.e. it will be stored on the default device, but allow implicit movement to other devices when necessary:
```

File: docs/JAX_for_PyTorch_users.md [hunk 4]
```diff
@@ -0,0 +1,890 @@
...
+
+```{code-cell} ipython3
+x = jnp.ones((3, 4))
+
+x.device, (x_cpu + x).device
+```
+
+However, if we make a computation with two arrays with explicitly specified devices, e.g. CPU and CUDA, similarlyto PyTorch, an error will be raised.
```
Comment (jakevdp) [inline docs/JAX_for_PyTorch_users.md:274]:
```suggestion
However, if we make a computation with two arrays with explicitly specified devices, e.g. CPU and CUDA, similarly to PyTorch, an error will be raised.
```

File: docs/JAX_for_PyTorch_users.md [hunk 5]
```diff
@@ -0,0 +1,890 @@
...
+```{code-cell} ipython3
+x = jnp.ones((3, 4))
+x_cpu = jax.device_put(x, device=jax.devices("cpu")[0])
+x_cuda = jax.device_put(x_cpu, device=jax.devices("cuda")[0])
+print(f"{x.device} -> {x_cpu.device} -> {x_cuda.device}")
+```
+
+```{code-cell} ipython3
+
+```
```
Comment (jakevdp) [inline docs/JAX_for_PyTorch_users.md:294]: Remove empty code cell?

File: docs/JAX_for_PyTorch_users.md [hunk 6]
```diff
@@ -0,0 +1,890 @@
...
+  base_uri: https://localhost:8080/
+id: icgYp1wKYnGi
+outputId: a58d0290-145a-4cd0-d78b-10c4729c0bc7
+---
+agg = tensor.sum()
+agg_value = agg.item()
+print(agg_value, isinstance(agg_value, float), isinstance(agg, jax.Array))
+```
+
+JAX is following NumPy's lead and recommending a functional approach and thus JAX Array API exposes less array methods compared to PyTorch Tensors:
```
Comment (jakevdp) [inline docs/JAX_for_PyTorch_users.md:386]: I know I suggested this, but I think it's not sounding right. How about:

> JAX follows NumPy in exposing a number of reduction and other operations as array methods:

File: docs/JAX_for_PyTorch_users.md [hunk 7]
```diff
@@ -0,0 +1,890 @@
...
+```python
+# Commonly exposed class ops
+jax_array = jnp.ones((2, 3))
+jax_array.sum(), jax_array.mean(), jax_array.min(), jax_array.max(), jax_array.dot(jax_array), # ...
+
+tensor = torch.ones(2, 3)
+tensor.sum(), tensor.mean(), tensor.min(), tensor.max(), tensor.matmul(tensor.T), # ...
+
```
Comment (jakevdp) [inline docs/JAX_for_PyTorch_users.md:394]: Then break this to another cell and say "PyTorch exposes many more methods on its tensor object than either JAX or NumPy does on their respective array objects. Here are some examples of methods only available in PyTorch:

File: docs/JAX_for_PyTorch_users.md [hunk 8]
```diff
@@ -0,0 +1,890 @@
...
+# PyTorch only exposed class ops
+tensor.sigmoid(), tensor.softmax(dim=1), tensor.sin(), # ...
+```
+
+```{code-cell} ipython3
+
```
Comment (jakevdp) [inline docs/JAX_for_PyTorch_users.md:400]: Remove empty cell

File: docs/JAX_for_PyTorch_users.md [hunk 9]
```diff
@@ -0,0 +1,890 @@
...
+Other AI/ML tutorials to check out:
+- [JAX AI Stack tutorials](https://jax-ai-stack.readthedocs.io/en/latest/tutorials.html)
+
+```{code-cell} ipython3
+
+```
+
+```{code-cell} ipython3
+
+```
```
Comment (jakevdp) [inline docs/JAX_for_PyTorch_users.md:838]: Remove empty code cells

File: docs/JAX_for_PyTorch_users.md [hunk 10]
```diff
@@ -0,0 +1,890 @@
...
+## Just-In-Time (JIT) compilation in JAX
+
+
+PyTorch users know very well about the eager mode execution of the operations in PyTorch, e.g. the operations are executed one by one without any high-level optimizations on sets of operations. Similarly, almost everywhere in this tutorial we used JAX in the eager mode as well.
+
+
+In PyTorch 1.0 [TorchScript](https://pytorch.org/docs/stable/jit.html) was introduced to optimize and serialize PyTorch models by capturing the execution graph into TorchScript programs, which can then be run independently from Python, e.g. as a C++ program.
+
+
+In JAX, there is a similar transformation: [`jax.jit`](https://jax.readthedocs.io/en/latest/_autosummary/jax.jit.html#jax.jit). It performs JIT compilation of a Python function for efficient execution in XLA. Behind the scenes, `jax.jit` is tracing the function wrapping the inputs into tracers to record all JAX operations. By default, JAX JIT is compiling the function on the first call and reusing the cached compiled XLA code on subsequent calls.
```
Comment (jakevdp) [inline docs/JAX_for_PyTorch_users.md:849]: "tracing the function wrappoing the inputs into tracers" sounds strange here. Is there a missing "and"?

File: docs/JAX_for_PyTorch_users.md [hunk 11]
```diff
@@ -0,0 +1,890 @@
...
+# Warm-up: compile the function
+_ = jit_matmul_relu(x, y)
+```
+
+```{code-cell} ipython3
+%%timeit
+jit_matmul_relu(x, y)
+```
+
+```{code-cell} ipython3
```
Comment (jakevdp) [inline docs/JAX_for_PyTorch_users.md:880]: Remove empty cell

File: docs/JAX_for_PyTorch_users.md [hunk 12]
```diff
@@ -0,0 +1,890 @@
...
+
+```
+
+### Further reading
+
+- https://jax.readthedocs.io/en/latest/jit-compilation.html
+
+```{code-cell} ipython3
```
Comment (jakevdp) [inline docs/JAX_for_PyTorch_users.md:888]: Remove empty cell

**Comment by vfdev-5** [comment]
Thanks for the review, @jakevdp, I applied all your suggestions

File: docs/conf.py [hunk 13]
```diff
@@ -55,4 +56,6 @@
 
 # List of patterns, relative to source directory, that match notebook
 # files that will not be executed.
-nb_execution_excludepatterns = []
+nb_execution_excludepatterns = [
+    'JAX_for_PyTorch_users.ipynb',
```
Comment (jakevdp) [inline docs/conf.py:60]: Skipping execution here means that the code in this tutorial will not have any test coverage. I understand this is probably because you showed GPU-specific execution, and we don't have GPUs in our readthedocs build environment. But I wonder if we could structure the discussion in a way that allows for notebook execution – for example mention CUDA support via unexectued code blocks rather than in code cells? What do you think?

**Comment by jakevdp** [comment]
You have a jupytext version mismatch – you can fix that locally by running `pre-commit`:
```
pip install pre-commit
pre-commit run jupytext --all-files
```
This will always use the correct version of jupytext

File: docs/conf.py (hunk 13 above)
Comment (vfdev-5) [reply docs/conf.py:60]: If we would like to execute this notebook on cpu-only hardware we would need to do the following changes:
- put cells using "cuda" into python code in the text
- add torch, torchvision to requirements.txt
- reduce training time of resnet18 on CIFAR10, 3 epochs -> 1 epoch and 100 iterations per epoch -> 10 iterations per epoch.

Up to you to tell me if we still would like to do that.

**Comment by jakevdp** [comment]
Looks good! Let's land and iterate.

**Review by jakevdp — APPROVED** [review]
